presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNThroughOuterJoin.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Range;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.TopNNode;
+
+import java.util.List;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.optimizations.QueryCardinalityUtil.extractCardinality;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.FULL;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.LEFT;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.RIGHT;
+import static io.prestosql.sql.planner.plan.Patterns.Join.type;
+import static io.prestosql.sql.planner.plan.Patterns.join;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+import static io.prestosql.sql.planner.plan.TopNNode.Step.PARTIAL;
+
+/**
+ * Transforms:
+ * <pre>
+ * - TopN
+ *    - Join (left, right or full)
+ *       - left source
+ *       - right source
+ * </pre>
+ * Into:
+ * <pre>
+ * - TopN
+ *    - Join
+ *       - TopN (present if Join is left or outer, not already limited, and orderBy symbols come from left source)
+ *          - left source
+ *       - TopN (present if Join is right or outer, not already limited, and orderBy symbols come from right source)
+ *          - right source
+ * </pre>
+ */
+public class PushTopNThroughOuterJoin
+        implements Rule<TopNNode>
+{
+    private static final Capture<JoinNode> JOIN_CHILD = newCapture();
+
+    private static final Pattern<TopNNode> PATTERN =
+            topN().with(source().matching(
+                    join().capturedAs(JOIN_CHILD).with(type().matching(type -> isLeftOrFullOuter(type) || isRightOrFullOuter(type)))));
```

Comment:
I think this expression makes it harder to reason about what this is doing that it should: `isLeftOrFullOuter(type) || isRightOrFullOuter(type)`. It's really just saying "is it an outer join?". So, I'd just replace this with: `type == LEFT || type == RIGHT || type == FULL`